### PR TITLE
PBL-34678 Improvements to CloudPebble API

### PR DESCRIPTION
**Pull request by Spacerat**
This PR represents a large change to the way that XHR is done between the CloudPebble client and server.

## Client Side: ES6-promisifying everything

All promises/deferreds and async callbacks are converted to ES6-compliant promises (specifically, the [Bluebird](http://bluebirdjs.com/docs/getting-started.html) library) without changing _too much_ else. This includes

- Making wrapper functions around `$.ajax`/`$.post`/`$.getJSON` which return real promises, and using the wrapper functions (`Ajax.Ajax`/`Ajax.Post`/`Ajax.Get`) everywhere instead.
- Handling `if (data.success == false)` in the wrapper function so that it rejects the promise. 
- Refactoring all AJAX calls to use `promise.catch()` instead of `if (data.success == false)`, so that all error handlers now catch HTTP errors (e.g. Internal Server Errors) and JSON errors in a uniform way.
- Changing many functions which used callbacks to use promises instead. This allows them to chain better with everything else and further unifies error handling.
-  Simplifying and refactoring code in a number of places where Deferreds were being created unnecessarily and just resolved or rejected with the result of another promise. In many cases, a function can just directly return the promise created by a call to `Ajax.Post`.

## Server side: @json_view decorator

All JSON API functions now use a decorator which ensures a few important things:

- The output from the view is always JSON, even when there are unhandled exceptions
- All errors produce HTTP error codes.
- Exception tracebacks will always be printed

Other features:

- Functions can now throw "InternalServerError" or "BadRequest" exceptions to explicitly return 400 or 500 error codes.
- In debug mode, tracebacks are attached to the JSON error object so that the CloudPebble client can print them.

## Testing

Since these changes effects practically every aspect of CloudPebble, everything needs to be re-tested. The following is a checklist of things tested (or not tested) so far.

Projects home

- [x]  Import a zip project
- [x]  Import a github project
- [x]  Change user settings
- [x]  Import Gist
- [x]  Link to github

Timeline

- [x]  View timeline (preview) page
- [ ]  Insert a pin
- [ ]  Delete a pin

Github Project Integration

- [x]  Push
- [x]  Pull

Source file

- [x]  Create a new source file
- [x]  Open a source file
- [x]  Edit/save a source file
- [x]  Rename a source file
- [x]  Delete a source file
- [x]  Reload a source file
- [x]  Show YCM autocompletions
- [x]  Show YCM errors
- [x]  Use cmd-click goto

Fuzzyprompt

- [x]  Go to (a specific line in) source file
- [x]  Go to a resource
- [x]  Go to all the other pages

Resource file

- [x]  Open a resource file
- [x]  Create a new resource
- [x]  Delete a resource
- [x]  Edit a resource’s properties
- [x]  Upload new files for a resource
- [x]  Add an identifier to a resource

Project settings

- [x]  Change project settings

Building

- [x]  Start building a project
- [x]  Poll a project build for status
- [x]  Show build failure for a broken project
- [x]  Show build success for a successful project
- [x]  Download a PBW
- [x]  View a build log
- [x]  Download project as ZIP

Shortcuts:

- [x]  Save with ctrl-s
- [x]  Save-all with ctrl-shift-s
Misc:

- [x]  saveall/compile/run button

Emulator

- [x]  Run a project in QEMU
- [x]  Ping/kill the emulator
- [x]  Press emulator buttons

### CODE REVIEW

File: ide/api/phone.py [hunk 1]
```diff
@@ -22,14 +23,18 @@ def list_phones(request):
         params={'client_id': settings.SOCIAL_AUTH_PEBBLE_KEY})
 
     if response.status_code != 200:
-        return json_failure(response.reason)
+        if 400 <= response.status_code < 500:
```
Comment (Katharine) [inline ide/api/phone.py]: Why are you making this distinction? It's not the client that made a bad request here.

Comment (Spacerat) [inline ide/api/phone.py]: If I remember correctly, I made the decision to make it show you the message for some particular errors but not others, although I can't remember exactly what the errors were. I could just raise an InternalServerError or Exception in all cases if you prefer.

Comment (Katharine) [inline ide/api/phone.py]: This code has been completely unused for almost a year, so it probably doesn't actually matter.

File: ide/static/ide/js/compile.js [hunk 2]
```diff
@@ -76,36 +72,45 @@ CloudPebble.Compile = (function() {
...
+            return Ajax.Get('/ide/project/' + PROJECT_ID + '/build/history').then(function(data) {
+                CloudPebble.ProgressBar.Hide();
+                pane.removeClass('hide');
+                if (data.builds.length > 0) {
+                    update_last_build(pane, data.builds[0]);
+                } else {
+                    update_last_build(pane, null);
+                }
+                pane.find('#' +
+                    'run-build-table').html('');
```
Comment (Katharine) [inline ide/static/ide/js/compile.js]: I am confused by this line wrapping.

File: ide/static/ide/js/editor.js [hunk 3]
```diff
@@ -15,748 +15,716 @@ CloudPebble.Editor = (function() {
...
-                };
-
-                var ib_pane = $('#ui-editor-pane-template').clone().removeClass('hide').appendTo(pane).hide();
-                var ib_editor = new IB(ib_pane.find('.ui-canvas'), ib_pane.find('#ui-properties'), ib_pane.find('#ui-toolkit'), ib_pane.find('#ui-layer-list > div'));
-                var ib_showing = false;
-
-                CloudPebble.GlobalShortcuts.SetShortcutHandlers({
-                    save: save
+                }).catch(function(error) {
+                    alert(gettext("Failed to reload file. ") + error);
```
Comment (Katharine) [inline ide/static/ide/js/editor.js]: You should be using interpolation, not concatenation.

File: ide/static/ide/js/compile.js (hunk 2 above)
Comment (Spacerat) [inline ide/static/ide/js/compile.js]: Oops. That was either pycharm or a typo.

File: root/static/common/js/ajax.js [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+                    if (_.isObject(data) && !!success_key && !data[success_key]) {
+                        handle_traceback(data.traceback);
+                        reject(data.error ? new Error(data.error) : new Error(gettext("Unknown error")));
+                    }
+                    else {
+                        resolve(data);
+                    }
+                }).fail(function (jqXHR, textStatus, errorThrown) {
+                    if (textStatus == 'abort') {
+                        resolve();
```
Comment (Katharine) [inline root/static/common/js/ajax.js]: Why?

File: root/static/common/js/ajax.js [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+                    }
+                    var message;
+                    if (jqXHR && jqXHR.responseJSON && jqXHR.responseJSON.error) {
+                        message = jqXHR.responseJSON.error;
+                        handle_traceback(jqXHR.responseJSON.traceback);
+                    }
+                    else {
+                        message = errorThrown;
```
Comment (Katharine) [inline root/static/common/js/ajax.js]: If this naming is correct, then you end up doing Error(Error(message)), which is probably not the intention. If it's not, it should be renamed.

File: root/static/common/js/ajax.js (hunk 4 above)
Comment (Spacerat) [inline root/static/common/js/ajax.js]: aborts are intentional, so it doesn't really make sense to treat them as promise failures. That being said, I put this in specifically for some feature in the test-bench branch which would cancel requests and that no longer exists. I don't think CloudPebble cancels requests anywhere else, so I could get rid of this check.

Comment (Katharine) [inline root/static/common/js/ajax.js]: I'd remove it; it's a bit bizarre.

File: root/static/common/js/ajax.js (hunk 5 above)
Comment (Spacerat) [inline root/static/common/js/ajax.js]: I'm not quite sure what you mean by this. There's only one layer of new Error() here that I can see?

Comment (Katharine) [inline root/static/common/js/ajax.js]: The name `errorThrown` implies that it is itself an `Error`. You then do `new Error(errorThrown)`.

Comment (Spacerat) [inline root/static/common/js/ajax.js]: Ah right I see. No, errorThrown is a jQuery Ajax thing. It's just a string like "Not Found" or "Internal Server Error" which gets set when a request fails. Here I'm just setting the message to be that if the response didn't have an actual message itself.

Comment (Katharine) [inline root/static/common/js/ajax.js]: Well then, rename it. :p

**Pull request merged**